homeassistant/components/tplink/climate.py

```diff
@@ -0,0 +1,141 @@
+        | ClimateEntityFeature.TURN_OFF
+        | ClimateEntityFeature.TURN_ON
+    )
+    _attr_hvac_modes = [HVACMode.HEAT, HVACMode.OFF]
+    _attr_precision = PRECISION_WHOLE
+
+    # This disables the warning for async_turn_{on,off}, can be removed later.
+    _enable_turn_on_off_backwards_compatibility = False
+
+    def __init__(
+        self,
+        device: Device,
+        coordinator: TPLinkDataUpdateCoordinator,
+        *,
+        parent: Device,
+    ) -> None:
+        """Initialize the switch."""
+        super().__init__(device, coordinator, parent=parent)
+        self._state_feature = self._device.features["state"]
+        self._mode_feature = self._device.features["thermostat_mode"]
+        self._temp_feature = self._device.features["temperature"]
+        self._target_feature = self._device.features["target_temperature"]
+
+        self._attr_min_temp = self._target_feature.minimum_value
+        self._attr_max_temp = self._target_feature.maximum_value
+        self._attr_temperature_unit = UNIT_MAPPING[cast(str, self._temp_feature.unit)]
+
+    @async_refresh_after
+    async def async_set_temperature(self, **kwargs: Any) -> None:
+        """Set target temperature."""
+        if (temperature := kwargs.get(ATTR_TEMPERATURE)) is None:
+            return
+        await self._target_feature.set_value(int(temperature))
+
+    @async_refresh_after
+    async def async_set_hvac_mode(self, hvac_mode: HVACMode) -> None:
+        """Set hvac mode (on/off)."""
+        if hvac_mode is HVACMode.HEAT:
+            await self._state_feature.set_value(True)
+        elif hvac_mode is HVACMode.OFF:
+            await self._state_feature.set_value(False)
+        else:
+            _LOGGER.warning("Tried to set unsupported mode: %s", hvac_mode)
+
+    @async_refresh_after
+    async def async_turn_on(self) -> None:
+        """Turn heating on."""
+        await self._state_feature.set_value(True)
+
+    @async_refresh_after
+    async def async_turn_off(self) -> None:
+        """Turn heating off."""
+        await self._state_feature.set_value(False)
+
+    @callback
+    def _async_update_attrs(self) -> None:
+        """Update the entity's attributes."""
+        self._attr_current_temperature = self._temp_feature.value
+        self._attr_target_temperature = self._target_feature.value
+
+        self._attr_hvac_mode = (
+            HVACMode.HEAT if self._state_feature.value else HVACMode.OFF
+        )
+
+        STATE_TO_ACTION = {
+            ThermostatState.Idle: HVACAction.IDLE,
+            ThermostatState.Heating: HVACAction.HEATING,
+            ThermostatState.Off: HVACAction.OFF,
+        }
+        if (
+            self._mode_feature.value not in STATE_TO_ACTION
+            and self._attr_hvac_action is not HVACAction.OFF
+        ):
+            _LOGGER.warning(
+                "Unknown thermostat state, defaulting to OFF: %s",
+                self._mode_feature.value,
+            )
```

Comment:
It looks like this should raise `ServiceValidationError`

https://developers.home-assistant.io/docs/integration_quality_scale_index#silver-

Reply:
Sorry thats a bad suggestion on my part. I was reviewing on mobile and lost track of which function this is in.